### ISSUE TITLE
TFLiteConverter轉換問題

### ISSUE THREAD

**Issue by Hank880223** [alibaba/TinyNeuralNetwork#217]
請問TFLiteConverte只能用在量化後的模型嗎？
我用量化後的PyTorch模型轉換成TFLite模型後, 執行的結果是正常的
![out](https://user-images.githubusercontent.com/84907438/231825469-b510f2d8-6845-4d9a-84ab-175a831c6c74.jpg)

使用未量化的PyTorch模型轉成TFLite模型後, 執行的結果是錯誤的
![out (copy)](https://user-images.githubusercontent.com/84907438/231825510-fbbc2cb0-0943-4059-8155-0a9f05e4ac61.jpg)

**Comment by peterjc123** [comment]
@Hank880223 非量化模型你是如何转换的？我们支持浮点模型的转换啊

**Comment by Hank880223** [comment]
@peterjc123 我目前已經解決了, 我在轉非量化模型的過程中, 因為BN的eps和momentum的設定是模型建立完後才處理的, 忘記設定才導致上述的結果錯誤, 因此將BN的參數設定好後, 結果就正常了.
 
另外我想問, 我在Netron看到TFLiteConverte轉換後的模型中, DepthwiseConvolution2D在kernel size=3x3, stride=1的padding模式是same, 這個是TFLiteConverte中自動優化的嗎？那kernel size=3x3, stride=2是否也可以將pad融合進去變成same模式?

**Comment by peterjc123** [comment]
> DepthwiseConvolution2D在kernel size=3x3, stride=1的padding模式是same, 這個是TFLiteConverte中自動優化的嗎？

是的，我们里面有计算图优化器来完成这个事情。



> 那kernel size=3x3, stride=2是否也可以將pad融合進去變成same模式?

不行，具体可以看tensorflow的same padding模式的说明 https://www.tensorflow.org/api_docs/python/tf/nn#same_padding
我们基本是按照他的定义来做判断条件的 
https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/converter/operators/optimize.py#L200

**Comment by Hank880223** [comment]
@peterjc123 感謝你的幫忙, 大致上瞭解了

**Comment by peterjc123** [comment]
@Hank880223 
> 那kernel size=3x3, stride=2是否也可以將pad融合進去變成same模式?
我又看了下这个问题，感觉我好像理解错你的意思了。你是想问怎么做能翻译出kernel size=3x3, stride=2的same padding吧
由于这种情况下，做的是非对称padding，所以不能直接用`nn.Conv`下面的`padding`参数，而是需要对输入直接用`F.pad`来做padding。实际代码例子如下:

```py
import torch
from tinynn.converter import TFLiteConverter

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(3, 3, 3, 2)
    
    def forward(self, x):
        x = torch.nn.functional.pad(x, (0, 1, 0, 1))
        return self.conv(x)

model = Model()
model.cpu()
model.eval()

dummy_input = torch.rand((1, 3, 224, 224))

output_path = os.path.join(CURRENT_PATH, 'out', 'test.tflite')

converter = TFLiteConverter(model, dummy_input, output_path)
converter.convert()

```

**Comment by Hank880223** [comment]
@peterjc123 
> 你想問怎麼才能翻譯出kernel size=3x3, stride=2的same padding吧
是的, 感謝您提供的方法, 目前已經測試過了, 可以順利轉換成same padding了
